### ISSUE TITLE
indent with "4" or "2" spaces

### ISSUE THREAD

**Issue by leocaseiro** [buzinas/tslint-eslint-rules#74]
Hi, Thanks for this TSLINT extension. That would be nice if we replace the indent method with the [eslint one](http://eslint.org/docs/rules/indent).

So we could set the number of spaces, like:

`{indent: [true, 2]}`
`{indent: [true, 4]}`
`{indent: [true, "tab"]}`

What do you think?

Thanks

**Comment by buzinas** [comment]
Hello, @leocaseiro, thanks for the issue.

If we're going to support this, I won't reinvent the wheel, so we should have a separate rule for that, e.g:

```json
indent-size: [true, 2]
```

And then we could add to the docs to combine it with the `indent` rule, when using spaces.

If you have the time, take a look at [the Contributors Guide](https://github.com/buzinas/tslint-eslint-rules#contributing) and feel free to send a PR for this! :)

**Comment by buzinas** [comment]
@jmlopez-rod Would you be able to see this one?

**Comment by jmlopez-rod** [comment]
I would love to. I'm getting tired of seeing some of my project files without the proper level of indent.

**Comment by buzinas** [comment]
@jmlopez-rod I thank you in advance :)

**Comment by jmlopez-rod** [comment]
Just wanted to write an update. I believe I have successfully ported the rule from [eslint](https://github.com/eslint/eslint/blob/master/lib/rules/indent.js). I say I believe because I have manage to make all their [tests](https://github.com/eslint/eslint/blob/master/tests/lib/rules/indent.js) pass.

At this point I can't make the PR since I have lots to clean up. I need to properly format all the tests (they all use double quotes) and do an overall clean up of the code. Once this is done I hope @buzinas can do a code review.

Keep in mind that the rule won't be all complete since it doesn't cover typescript related stuff (decorators), so I'll need some help with test cases.

**Comment by jmlopez-rod** [comment]
Ok, this is almost done, I took it out in wild by including the js file in my other projects and it seems to work ok (found one tiny bug though). There's a bit more refactoring I'd like to do but for anyone that wants to start using it can go ahead and do what I did:

https://gist.github.com/jmlopez-rod/da74fd48e8f963322ba24a92947b5ed0

Just copy that file to `node_modules/tslint-eslint-rules/dist/rules/terIndentRule.js` and add the following to your configuration:

```
"ter-indent": [
      true, 2,
      {
        "FunctionDeclaration": {
          "parameters": 1,
          "body": 1
        },
        "FunctionExpression": {
          "parameters": 1,
          "body": 1
        },
        "MemberExpression": 1
      }
    ],
```

I would like to get the rule name resolve before I do the PR though.

**Comment by jmlopez-rod** [comment]
Does anyone who have tried the rule have any comments on the name? I'd like to make the PR sometime next weekend and start porting other rules. I'd really like to have the names to match with eslint as close as possible to avoid confusion.

**Comment by noxan** [comment]
I was able to find only a single issue - no, first of all great job, works like a charm! :)

But getting back on giving you some feedback. I had an issue with `undefined` values as parameters for the following two functions (line numbers ahead):

- 16: `function isKind(node, kind) {`
- 170: `IndentWalker.prototype.getNodeIndent = function (node) {`

You should be able to reproduce the issue with the following snippet:

```
import { Http } from '@angular/http';

class Demo {
  constructor(private http: Http) {}

  makeRequest() {
    return this.http.get('/')
    .map(res => res.json()); // <- this line kills it!
  }
}
```

My understanding of those linting internals is rather basic, but feel free to get back on me if I can assist you in any way. I would be glad to see this rule in one of the next releases!

**Comment by jmlopez-rod** [comment]
@noxan , thanks for the feedback. I think some of the latest changes I made may have fixed it. I just tried your example (will add it to the tests) and if I purposely do it wrong I get the expected results:

```
this.http.get('/')
 .map(res => res.json())
     -------

     Found 1 error(s):

       (2) Expected indentation of 2 spaces but found 1.
```

This is the configuration used:

```
[2, { MemberExpression: 1 }]
```
Yeah, I would also be happy if this makes it in soon along with max-line rule. I have modified the gist file with the compiled file so you may continue using while we get the release.

**Comment by leocaseiro** [comment]
Hi @jmlopez-rod, thank you so much for this rule. It's really nice.

However, I'm not sure why I'm having the wrong maths.

I set to be:

```javascript
        "ter-indent": [
            true, 4
        ],
```

So I expect to validate always 4 spaces for each indentation.

However, I'm getting the wrong messages.

This is my code, I have a switch case, inside a method, inside my class.
I have also `no-trailing-whitespace` enabled.

Eq:
```javascript
export class Datepicker {
....getFormat(type) {
........let format: string;

........switch (type) {
............case 'Time':
................format = 'HH:mm:ss';
................break;

............case 'DateTime':
................format = 'DD/MM/YYYY HH:mm:ss';
................break;

............default:
................format = 'DD/MM/YYYY';
................break;
........}

........return format;
....}
}
```

All my `case` indentations, I get the follow error:

> Expected indentation of 8 spaces but found 12.


Is that right? Am I doing something wrong? Thanks in advance

**Comment by leocaseiro** [comment]
Never mind, I've just found my issue.
By default, SwitchCase is not expected to be indented, so you need to enable, like so:

```javascript
"ter-indent": [
            true, 4, {"SwitchCase": true}
        ],
```

**Comment by leocaseiro** [comment]
@jmlopez-rod in regards to the name of the function, honestly I believe something like:
`indent-rules` or `indent-spaces` would be a better name.

I'm not sure where did you get the `ter` from.

Other than that. I'd love this PR. Even though I'm using it already in my source.

Very very happy.
Sorry for my delay.

**Comment by jmlopez-rod** [comment]
@leocaseiro , thank you for the feedback and I'm glad its already being used. In regards to the name please see this issue: https://github.com/buzinas/tslint-eslint-rules/issues/134

There are many names that a rule can have and many potential conflicts with rules that TSLint may bring. For that reason I would really like to have a prefix for future rules that come out of this project so that we can easily identify them as such and at the same time easily know the correspondent rule from ESLint.

It had been a while since I opened the issue so I thought it would be good to get the discussion going.

**Comment by leocaseiro** [comment]
Hi @jmlopez-rod, forgive me. I've just realised that `ter` stands for `tslint-eslint-rules`.

Keep it. Loved it! I'm happy to close the issue when gets merged :D

I've set up in 2 projects already and had some fixes to do!
AWESOME!

**Comment by tarruda** [comment]
> By default, SwitchCase is not expected to be indented, so you need to enable, like so:

@leocaseiro It seems interfaces are not checked, at least not by default. Do you know an option to enable it?

**Comment by jmlopez-rod** [comment]
@tarruda Seems that I never considered them. Could you open a new issue reporting this?

**Comment by tarruda** [comment]
> @tarruda Seems that I never considered them. Could you open a new issue reporting this?

Done: #158

**Comment by YvesBoah** [comment]
Il faudrait commenté ces lignes de codes  se situant dans le répertoire 
     build/webpack.base.conf.js
      // {
      //   test: /\.(js|vue)$/,
      //   loader: 'eslint-loader',
      //   enforce: 'pre',
      //   include: [resolve('src'), resolve('test')],
      //   options: {
      //     formatter: require('eslint-friendly-formatter')
      //   }
      // },
